pkg/crinit/aggregated/aggregated.go

```diff
@@ -0,0 +1,427 @@
+
+// NewCmdAggregated defines the `aggregated` command that bootstraps a cluster registry
+// inside a host Kubernetes cluster.
+func NewCmdAggregated(cmdOut io.Writer, pathOptions *clientcmd.PathOptions, defaultServerImage, defaultEtcdImage string) *cobra.Command {
+	opts := &aggregatedClusterRegistryOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "aggregated CLUSTER_REGISTRY_NAME --host-cluster-context=HOST_CONTEXT",
+		Short:   "Initializes an aggregated cluster registry",
+		Long:    aggregated_long,
+		Example: aggregated_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := opts.SetName(args)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = validateOptions(opts)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = opts.MarshalOptions()
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			hostConfig, err := util.GetClientConfig(pathOptions, opts.Host, opts.Kubeconfig).ClientConfig()
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+			hostClientset, err := client.NewForConfig(hostConfig)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+			apiServiceClientset, err := apiregclient.NewForConfig(hostConfig)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+			err = Run(opts, cmdOut, hostClientset, apiServiceClientset, pathOptions)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.BindCommon(flags, defaultServerImage, defaultEtcdImage)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// validateOptions ensures that options are valid.
+func validateOptions(opts *aggregatedClusterRegistryOptions) error {
+	opts.ApiServerServiceType = v1.ServiceType(opts.apiServerServiceTypeString)
+	return opts.ValidateCommonOptions()
+}
+
+// Run initializes a cluster registry.
+func Run(opts *aggregatedClusterRegistryOptions, cmdOut io.Writer,
+	hostClientset client.Interface, apiSvcClientset apiregclient.Interface, pathOptions *clientcmd.PathOptions) error {
+
+	err := opts.CreateNamespace(cmdOut, hostClientset)
+	if err != nil {
+		return err
+	}
+
+	svc, ips, hostnames, err := opts.CreateService(cmdOut, hostClientset)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := opts.GenerateCredentials(cmdOut, svc.Name, ips, hostnames,
+		false, false)
+	if err != nil {
+		return err
+	}
+
+	err = opts.CreateAPIServerCredentialsSecret(hostClientset, credentials)
+	if err != nil {
+		return err
+	}
+
+	pvc, err := opts.CreatePVC(cmdOut, hostClientset, svc.Name)
+	if err != nil {
+		return err
+	}
+
+	err = opts.CreateAPIServer(cmdOut, hostClientset, false, false, ips, pvc)
+	if err != nil {
+		return err
+	}
+
+	_, err = createAPIService(cmdOut, apiSvcClientset, opts,
+		util.GetCAKeyPair(credentials).Cert)
+	if err != nil {
+		return err
+	}
+
+	err = createRBACObjects(cmdOut, hostClientset, opts)
+	if err != nil {
+		return err
+	}
+
+	err = opts.UpdateKubeconfig(cmdOut, pathOptions, svc, ips, hostnames,
```

Comment:
Couldn't find the old comment thread here, so starting a new one.

I think you should remove this for now. It's easier to add it back later if people want this functionality, and I'm not sure why a user would want to interact with the cluster registry like this if they're aggregating it.

Sorry, I thought I had made that comment already in the last pass!

Reply:
Yeah I saw this comment but I couldn't reply to it.

I actually did try to remove it but ran into an error caused by the subsequent call to `WaitForAPIServer`. Since that method is shared by both standalone and aggregated I opted to worry about it later since it doesn't affect the functionality. So I don't disagree, but would prefer to follow up with another PR. WDYT?

Reply:
SGTM.